### PR TITLE
Fix bug when rounding large numbers to floating point types

### LINKED ISSUE #52355: `round(::Type{<:AbstractFloat}, x, ::RoundingMode)` violates docstring
*I don't have the development branch checked out right now, so forgive me for manually-interpreting this code and let me know if any of my concerns are due solely to my mistakes. This issue is in response to new functionality introduced in #50812, which has not been included in a release version (as of v1.9).*

I take issue with the implementation
```julia
round(::Type{T}, x, r::RoundingMode) where T = convert(T, round(x, r))
```
and the docstrings related to rounding functions. For example, `?floor` in #50812 states
> `floor(x)` returns the nearest integral value of the same type as `x` that is less than or
equal to `x`.
`floor(T, x)` converts the result to type `T`, throwing an `InexactError` if the floored
value is not representable a `T`.

(note the minor typo "not representable a `T`", but that's not why I'm here)

However, under the above implementation we have `floor(Float32, 0xffff_ffff) == 2f0^32 > 0xffff_ffff`. The `floor(0xffff_ffff)` part is executed correctly (no-op because integer argument), but the subsequent `convert(Float32, 0xffff_ffff)` results in upward rounding and yields an invalid result (being larger than the input). By the current docstring, the result should be an `InexactError` since `0xffff_ffff` is not representable as a `Float32`. Alternatively, one might suggest that the largest not-greater `Float32` be returned (in this case `prevfloat(2f0^32)`) but that would require a docstring change.

Possible solutions include:
- Change the docstring to remark that `convert` is naively applied to the result of `floor(x)`, so that this behavior is explainable under the stated semantics.
- Throw an `InexactError`, as the written docstring indicates should happen.
- Remove the unrestricted `Type` argument from the rounding functions. Retain some special cases like `Type{<:Integer}` (which already exist in release versions) where such issues cannot arise.
  - On current release v1.9, `floor(Float32, 0xffff_ffff)` is a `MethodError`
  - Is `floor(T, x)` really that much better than `T(floor(x))` or `convert(T, floor(x))`? I'd be happy to have it except that it's wrong here. At least with the latter two the return value is explainable via the rounding behavior of `T` conversion.
- Adjust the implementation to return `prevfloat(floor(T,x))` when `floor(T,x) > x` and `T<:AbstractFloat`, with similar changes to other rounding functions. Adjust the docstrings accordingly to accomodate this possibility. Something in the flavor of "return the largest integral-valued `T` that is less than or equal to `x`".
  - This might be somewhat expensive since Julia doesn't currently give control over hardware rounding modes (to get the correct result in the first place) and integer-float comparisons aren't cheap (to fix the result in post-processing).

### PR DESCRIPTION
**Pull request by LilithHafner**
- fix #52355 using option 4 (round to nearest representable integer)
- update docstrings *including documenting convert to Inf behavior even though Inf is not the "closest" floating point value*
- add some assorted tests

### CODE REVIEW

File: test/rounding.jl [hunk 1]
```diff
@@ -458,3 +458,15 @@ end
     @test_throws InexactError round(Int128, -Inf16)
     # More comprehensive testing is present in test/floatfuncs.jl
 end
+
+@test "floor(<:AbstractFloat, large_number) (#52355)" begin
+    @test floor(Float32, 0xffff_ffff) == prevfloat(floatmax(Float32)) <= 0xffff_ffff
```
Comment (mikmoore) [inline test/rounding.jl:463]:
```suggestion
    @test floor(Float32, 0xffff_ffff) == prevfloat(2f0^32) <= 0xffff_ffff
```
wrong value

File: base/float.jl [hunk 2]
```diff
@@ -464,6 +464,20 @@ round(x::IEEEFloat, ::RoundingMode{:Down})    = floor_llvm(x)
 round(x::IEEEFloat, ::RoundingMode{:Up})      = ceil_llvm(x)
 round(x::IEEEFloat, ::RoundingMode{:Nearest}) = rint_llvm(x)
 
+rounds_up(x, ::RoundingMode{:Down}) = false
+rounds_up(x, ::RoundingMode{:Up}) = true
+rounds_up(x, ::RoundingMode{:ToZero}) = x < 0
+rounds_up(x, ::RoundingMode{:FromZero}) = x > 0
```
Comment (mikmoore) [inline base/float.jl:470]:
```suggestion
rounds_up(x, ::RoundingMode{:ToZero}) = signbit(x)
rounds_up(x, ::RoundingMode{:FromZero}) = !signbit(x)
```
I suppose zero is never rounded, in which case I don't think this matters. But if these get used in some other functionality later, `signbit` may be the more robust option.

Comment (LilithHafner) [reply base/float.jl:470]: signbit is also probably faster in some cases.

**Review by mikmoore — APPROVED** [review]

**Comment by LilithHafner** [comment]
CI is very red, but all of the failures look very unrelated (internet or file system) and master has been red for a while.

**Comment by LilithHafner** [comment]
> I expect there are situations where an `InexactError` may still be thrown

I don't think so

```julia
julia> Base.infer_effects(round, Tuple{Type{Float64}, Int})
(+c,+e,+n,+t,+s,+m,+u)
```

> or could be for some other types

Yes:

```julia
julia> floor(UInt8, 1000)
ERROR: InexactError: trunc(UInt8, 1000)
```

We also have

>  floor(T, x) converts the result to type T, throwing an InexactError if the floored value is not representable a T.

in the docstring already.

---

> the current docstrings don't address the new behavior introduced in this PR.

Do you have a specific behavior that isn't documented? For the most part I think this PR is doing a bugfix to change behavior to match what was already quasi-documented.

**Comment by mikmoore** [comment]
Sorry. You're right. I just looked at this first thing in the morning and my brain wasn't working. It talks about rounding to representable values.

**Pull request merged**